source/manuals/programming-languages/python/python.html.md.erb

```diff
@@ -1,106 +1,150 @@
 
-    **Recommendation**: Put your top-level requirements in a
-    `requirements-app.txt` file and use [this Makefile
-    script][dm-deps-commit-makefile] to generate a fully specified
-    `requirements.txt`, which is used whenever you need to pip-install the
-    dependencies. The script creates a temporary virtualenv, pip-installs the
-    `requirements-app.txt` and pip-freezes the result as `requirements.txt`.
+A good strategy for specifying your application's Python dependencies has two desirable
+characteristics - they should be:
 
-    **Recommendation**: Use `requirements-dev.txt` for any dependencies used in
-    development/test but not production environments.
+1. Reproducible (predictable)
 
+    Pin your application's full dependencies – specific versions, rather than ranges – or you'll
+    get unpredictability between your dev environment and other environments. You want a
+    new starter to avoid small hard-to-spot problems. And you want parity between what you
+    test locally, what is tested by CI, and what you deploy, or you risk new issues appearing on
+    a live server. Additionally these things can be hard to diagnose.
 
 2. Kept up-to-date
 
-    Security issues are found in libraries, so it is important to choose
-    libraries that are maintained and to ensure your team has a strategy to
-    ensure security updates are installed without significant delay.
+    Security issues are found in libraries, so it is important to choose libraries that are
+    maintained and to ensure your team has a strategy to ensure security updates are installed
+    without significant delay. The [how to manage third party software dependencies][gds-way-deps] section
+    gives further context and discusses tools that can help, such as [snyk.io][].
+
+Your README should document an easy-to-follow process by which all your Python
+dependencies can be upgraded to get bug fixes and security fixes, without introducing
+breaking changes to your build.
+
+Your pinned dependencies should be fully specified in a file called `requirements.txt`, and checked
+into your version control system (VCS). For projects with only a small number of dependencies, maintaining
+this manually (for example, installing with `pip install`, then using `pip freeze`) may be adequate.
+
+For larger projects, GDS recommends the following approach, which was developed after various issues
+were found in current mainstream tooling (see [this commit message][dm-deps-commit], and note also the
+[lack of support for specifying VCS dependencies](https://github.com/jazzband/pip-tools/issues/355)
+in `pip-compile`).
+
+Put your top-level requirements in a `requirements-app.txt` file.
+
+Use a "freeze" script (as seen in [this Makefile][dm-deps-commit-makefile]) to generate the
+fully specified `requirements.txt`, which is used whenever you need to pip-install the
+dependencies.
+
+* The script creates a temporary virtualenv, pip-installs the `requirements-app.txt` and
```

Comment:
This feels like information that should belong in a readme.

Can we package up the makefile solution into a standalone repo so we don't have to copy and paste it across projects? Even if it was just a template, it seems like it would be simpler to link to that than use Digital Marketplace as the template and have to update this if they change anything.

Reply:
> Can we package up the makefile solution into a standalone repo so we don't have to copy and paste it across projects? Even if it was just a template, it seems like it would be simpler to link to a standalone repo than use Digital Marketplace as the template and have to update this if they change anything.

I'm sure this is something we could do, although it would have to be with the understanding that someone would have to maintain it.

Reply:
I think the important thing we're trying to communicate here are the steps one should take to `freeze` their requirements. I don't think it would be too much to put in the actual steps a dev should take, something like:

>>>
To get a requirements.txt with the pinned dependencies and sub-dependencies for your project you should:
1. Put your known, first level dependencies in a requirements-app.txt file
2. Create and activate a new, clean, virtualenv
3. `pip install -r` your `requirements-app.txt` file
4. Run `pip freeze`

At this step `pip freeze` will output everything currently installed in your new virtualenv. The dependencies you specified in your `requirements-app.txt` and any sub-dependencies (dependencies of dependencies). This is the pinned list of requirements that should go in your `requirements.txt`

5. Clean everything up by deactivating and `rm -rf`ing the new virtualenv.

Reply:
That way we don't need to include the part below about the specific drawbacks of our makefile (also the makefile version doesn't allow range-pinned dependencies whereas the above process does)

Reply:
Yeah I think this is a good idea, we'd build on @benvand's thing. I'd probably like it if he go it into production for at least one DM system first, though :-)

The reason the explanation is here is because we aren't really mandating _that_ Makefile, just the effect it produces, so you need (here) to know what it does if you are to reproduce that (via copy-paste or by building your own or using `pip-compile` or whatever.

(I've thought before about maintaining a simple template project, but I tend to the view that if you're starting something new, then you might prefer to read the GDS Way and then see some pointers to the existing Python projects that do this consistently/well. But sticking the script in a lib you can easily install would certainly be nice, for as long as we do something better than `pip-compile` does at least.)

Reply:
Sorry just to clarify, I was responding to the original post, not to @benvand...

I don't actually agree with ben on this so I think maybe we should take this to #Python :)